### PR TITLE
Add schema v2 bipartite adapters

### PR DESCRIPTION
**Pull request by anmorgunov**
why: continue the schema v2 adapter migration for providers that use the same molecule/reaction bipartite tree shape. Keeping these together lets the shared traversal logic be reviewed with all of its immediate consumers, without mixing in unrelated plain-tree or route-string formats.

what changed:
- Added v2 adapters for AiZynthFinder, SynPlanner, and Syntheseus.
- Added `build_bipartite_molecule(...)` for the shared molecule -> reaction -> molecule tree traversal.
- Preserved provider-specific reaction metadata: AiZynthFinder/Syntheseus metadata and SynPlanner mapped reaction SMILES.
- Added adapter-boundary contract tests for extraction, casting, target mismatch, invalid SMILES pruning, cycles, duplicate leaves, and non-tree molecule nodes.

risk:
- The main invariant is that a molecule node maps to at most one child reaction in schema v2 route trees. Multiple child reactions now fail with `adapter.route_not_tree` instead of silently picking one.
- SynPlanner canonicalizes with `remove_mapping=True` because its raw reaction/molecule strings may carry atom mapping.
- This is additive v2 adapter code; legacy adapters and workflow wiring are not changed.

verification:
- `uv run pytest tests/v2/adapters/test_aizynth.py tests/v2/adapters/test_synplanner.py tests/v2/adapters/test_syntheseus.py -q`
- `uv run pytest tests/v2/adapters -q`
- `uv run ruff check src/retrocast/v2/adapters/common.py src/retrocast/v2/adapters/aizynth.py src/retrocast/v2/adapters/synplanner.py src/retrocast/v2/adapters/syntheseus.py src/retrocast/v2/adapters/__init__.py tests/v2/adapters/test_aizynth.py tests/v2/adapters/test_synplanner.py tests/v2/adapters/test_syntheseus.py`
- `uv run ty check src/retrocast/v2/adapters/common.py src/retrocast/v2/adapters/aizynth.py src/retrocast/v2/adapters/synplanner.py src/retrocast/v2/adapters/syntheseus.py tests/v2/adapters/test_aizynth.py tests/v2/adapters/test_synplanner.py tests/v2/adapters/test_syntheseus.py`

follow-ups:
- DirectMultiStep, MolBuilder, MultiStepTTL, RetroChimera, SynLlama, and URSA remain in the migration stash for later focused PRs.

---
<a href="https://app.blacksmith.sh/ischemist/codesmith/project-procrustes/pr/105"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1782573554&installation_id=125602297&pr_number=105&repository=ischemist%2Fproject-procrustes&return_to=https%3A%2F%2Fgithub.com%2Fischemist%2Fproject-procrustes%2Fpull%2F105&signature=3ce0d5b260e2bbfdab366e5ebfd689a5f7b7f587e2bab7c78b52c7219c50338a"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **New Features**
  * Added support for converting synthesis routes from three additional planning tools: AiZynthFinder, SynPlanner, and Syntheseus into the internal Route format.
  * Each adapter validates route payloads, preserves metadata and annotations, and enforces target SMILES consistency.

* **Tests**
  * Comprehensive test coverage for all new adapters, including validation, casting, cycle detection, and metadata preservation scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ischemist/project-procrustes/pull/105?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<h3>Greptile Summary</h3>

This PR adds schema v2 adapters for AiZynthFinder, SynPlanner, and Syntheseus — three synthesis planning tools that share the same bipartite (mol → reaction → mol) tree shape. A shared `build_bipartite_molecule` traversal helper in `common.py` drives all three, with each adapter supplying a `reaction_fields` callback for provider-specific metadata.

- `common.py` gains `build_bipartite_molecule` / `_build_bipartite_molecule` with consistent cycle detection (path-based `visited` set), SMILES prune/strict mode, and structural invariant checks; the `visited` accumulator is correctly hidden behind the private helper, addressing the concern from the previous review.
- Three new adapters (`aizynth.py`, `synplanner.py`, `syntheseus.py`) and their corresponding test suites are additive; `synplanner` canonicalizes mol nodes with `remove_mapping=True` to handle atom-mapped molecule SMILES from that tool's output.
- `test_askcos.py` gains three new edge-case tests (missing UUID, missing chemical/reaction node, all-reactants-pruned) for consistency with the new adapter contract suite.

<h3>Confidence Score: 5/5</h3>

This PR is additive adapter code with no changes to existing adapters or workflow wiring; safe to merge.

The shared bipartite traversal in common.py is correct: cycle detection is path-based, prune/strict mode is consistently applied, and structural invariants raise unconditionally as documented. The public/private split on build_bipartite_molecule addressed the concern from the previous review. Test coverage is thorough across all adapters.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/retrocast/v2/adapters/common.py | Adds build_bipartite_molecule with correct path-based cycle detection, prune/strict mode, and structural validation; public/private split addresses previous review comment. |
| src/retrocast/v2/adapters/aizynth.py | New AiZynthFinder adapter: delegates traversal to build_bipartite_molecule, extracts mapped_reaction_smiles/template from metadata, stores full metadata as annotations. |
| src/retrocast/v2/adapters/synplanner.py | New SynPlanner adapter: uses remove_mapping=True throughout for molecule SMILES; stores the raw reaction-node SMILES as mapped_reaction_smiles (intended — SynPlanner emits atom-mapped reaction strings). |
| src/retrocast/v2/adapters/syntheseus.py | New Syntheseus adapter: mirrors AiZynthFinder structure; conditionally extracts mapped_reaction_smiles and template from metadata while always passing full metadata as annotations. |
| tests/v2/adapters/test_aizynth.py | Comprehensive contract tests covering extraction, casting, target mismatch, cycle detection, duplicate leaves, prune/strict modes, and structural node-type errors. |
| tests/v2/adapters/test_synplanner.py | Contract tests for SynPlanner including atom-mapped reaction SMILES annotation preservation and canonicalization of mapped molecule SMILES. |
| tests/v2/adapters/test_syntheseus.py | Contract tests for Syntheseus mirroring the AiZynthFinder suite; metadata preservation verified for template and mapped_reaction_smiles fields. |
| tests/v2/adapters/test_askcos.py | Adds missing-UUID, missing-node, and all-reactants-pruned edge-case tests for the existing ASKCOS adapter for consistency with the new contract suite. |
| src/retrocast/v2/adapters/__init__.py | Registers AiZynthFinderAdapter, SynPlannerAdapter, and SyntheseusAdapter in the public API; __all__ is updated correctly. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["build_bipartite_molecule(node)"] --> B["_build_bipartite_molecule(node, visited=set())"]
    B --> C{"node.type == mol?"}
    C -- No --> ERR1["raise adapter.node_type_invalid"]
    C -- Yes --> D["canonicalize_smiles(node.smiles)"]
    D -- InvalidSmilesError --> E{"mode == prune?"}
    E -- Yes --> RN["return None"]
    E -- No --> ERR2["re-raise"]
    D -- OK --> F{"canon_smiles in visited?"}
    F -- Yes --> ERR3["raise adapter.cycle_detected"]
    F -- No --> G{"in_stock=True OR no children?"}
    G -- Yes --> L["return Molecule - leaf"]
    G -- No --> H{"len(children) > 1?"}
    H -- Yes --> ERR4["raise adapter.route_not_tree"]
    H -- No --> I["reaction_node = children[0]"]
    I --> J{"reaction_node.type == reaction?"}
    J -- No --> ERR5["raise adapter.node_type_invalid"]
    J -- Yes --> K["recurse each reaction child with visited union canon_smiles"]
    K --> M{"any reactants built?"}
    M -- No, strict --> ERR7["raise adapter.reaction_empty"]
    M -- No, prune --> RN
    M -- Yes --> N["call reaction_fields(reaction_node)"]
    N --> O["return Molecule with product_of Reaction"]
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `src/retrocast/v2/adapters/common.py`, line 199-204 ([link](https://github.com/ischemist/project-procrustes/blob/dee7189d3f9269bca46eefdb8bbbfa729810f3a0/src/retrocast/v2/adapters/common.py#L199-L204)) 

   <a href="#">P2</a> **`adapter.route_not_tree` is always raised regardless of `mode`**

   In `prune` mode every SMILES error is silently swallowed, but a molecule node with multiple child reactions raises unconditionally. If the intent is that structural invariants (non-tree shape) should always be hard errors this is fine, but it's worth documenting explicitly — especially because a caller who sets `mode="prune"` expecting all malformed nodes to be quietly skipped will be surprised when this raises instead.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/retrocast/v2/adapters/common.py
   Line: 199-204

   Comment:
   **`adapter.route_not_tree` is always raised regardless of `mode`**

   In `prune` mode every SMILES error is silently swallowed, but a molecule node with multiple child reactions raises unconditionally. If the intent is that structural invariants (non-tree shape) should always be hard errors this is fine, but it's worth documenting explicitly — especially because a caller who sets `mode="prune"` expecting all malformed nodes to be quietly skipped will be surprised when this raises instead.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!
</details>

<sub>Reviews (3): Last reviewed commit: ["Cover v2 adapter edge cases"](https://github.com/ischemist/project-procrustes/commit/8828bf53af76207a99aa97bd89ec5c0e2ad5c21f) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=34386064)</sub>

### CODE REVIEW

File: src/retrocast/v2/adapters/synplanner.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                code="adapter.target_pruned",
+                context={"adapter": "synplanner", "target_id": target_id},
+            )
+        if target is not None:
+            expected_smiles = canonicalize_smiles(target.smiles, remove_mapping=True)
+            if route_target.smiles != expected_smiles:
+                raise adapter_target_mismatch(
+                    "synplanner", target.id, expected_smiles=expected_smiles, actual_smiles=route_target.smiles
+                )
+        return Route(target=route_target)
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply src/retrocast/v2/adapters/synplanner.py:78]: no, this is actually worse coding style to have such a large separation between try and except, imo. The ValidationError highlights any issues with corrupted payloads in raw_route. Construction errors should be handled by the helpers, we shouldn't create umbrella try except blocks

**Comment by anmorgunov** [comment]
Addressed the Greptile comment in `a8bce86` by keeping `build_bipartite_molecule(...)` as the public entry point and moving recursion state into private `_build_bipartite_molecule(...)`.

I intentionally did not apply the Gemini suggestion to catch `ValidationError` around the full `cast` body. The current narrower handling keeps error semantics truthful: raw input validation becomes `AdapterSchemaError`, chemistry failures stay chemistry failures, and adapter/model construction bugs do not get mislabeled as bad raw schema.

Verification:
- `uv run pytest tests/v2/adapters/test_aizynth.py tests/v2/adapters/test_synplanner.py tests/v2/adapters/test_syntheseus.py -q`
- `uv run ruff check src/retrocast/v2/adapters/common.py src/retrocast/v2/adapters/aizynth.py src/retrocast/v2/adapters/synplanner.py src/retrocast/v2/adapters/syntheseus.py tests/v2/adapters/test_aizynth.py tests/v2/adapters/test_synplanner.py tests/v2/adapters/test_syntheseus.py`
- `uv run ty check src/retrocast/v2/adapters/common.py src/retrocast/v2/adapters/aizynth.py src/retrocast/v2/adapters/synplanner.py src/retrocast/v2/adapters/syntheseus.py tests/v2/adapters/test_aizynth.py tests/v2/adapters/test_synplanner.py tests/v2/adapters/test_syntheseus.py`

File: src/retrocast/v2/adapters/synplanner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anmorgunov) [reply src/retrocast/v2/adapters/synplanner.py:78]: @gemini-code-assist idk if you can remember stuff, but this suggestion sucks

**Comment by anmorgunov** [comment]
Added behavior-level coverage for the meaningful uncovered v2 adapter branches in `8828bf5`.

Covered:
- full prune-to-`adapter.target_pruned` for AiZynthFinder, SynPlanner, Syntheseus, and ASKCOS
- strict empty-reaction handling through the bipartite adapter path
- malformed bipartite topology through an adapter: molecule child under molecule and reaction child under reaction
- ASKCOS missing backing UUID/node lookups for chemicals and reactions

Remaining uncovered lines after `uv run pytest tests/v2/adapters --cov=src/retrocast/v2/adapters --cov-report=term-missing -q`:
- `askcos.py:47`: malformed metadata template fallback, intentionally low-risk
- `common.py:107-108`: direct private-helper misuse with a non-molecule root; public adapters validate this before the helper
- `paroutes.py:271,274`: strict invalid-SMILES/cycle behavior is already covered, this is coverage line accounting rather than missing behavior

Verification:
- `uv run pytest tests/v2/adapters/test_aizynth.py tests/v2/adapters/test_synplanner.py tests/v2/adapters/test_syntheseus.py tests/v2/adapters/test_askcos.py -q`
- `uv run pytest tests/v2/adapters -q`
- `uv run pytest tests/v2/adapters --cov=src/retrocast/v2/adapters --cov-report=term-missing -q` reports 99% for v2 adapters
- `uv run ruff check tests/v2/adapters/test_aizynth.py tests/v2/adapters/test_synplanner.py tests/v2/adapters/test_syntheseus.py tests/v2/adapters/test_askcos.py`
- `uv run ty check tests/v2/adapters/test_aizynth.py tests/v2/adapters/test_synplanner.py tests/v2/adapters/test_syntheseus.py tests/v2/adapters/test_askcos.py`
